### ISSUE TITLE
x:DataType doesn't work with nested types from other assemblies

### ISSUE THREAD

**Issue by Sergio0694** [microsoft/microsoft-ui-xaml#3238]
**Describe the bug**

Using `x:DataType` doesn't work with nested types from other assemblies.

Consider the following viewmodel code:

```csharp
public class RootViewModel
{
    public IEnumerable<object> Sections { get; } = new[] { new NestedViewModel() }; 

    public class NestedViewModel
    {
        public string SampleText => "Hello world";
    }
}
```

**NOTE:** this code _must_ be placed in a separate project (eg. a .NET Standard 2.0 lib) to repro the issue.
It works fine if the classes are right in the UWP app project.

And the following XAML code:

```xml
<Page
    x:Class="BindToNestedTypeRepro.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:libs="using:BindToNestedTypeReproLibs">
    <Page.DataContext>
        <libs:RootViewModel x:Name="ViewModel"/>
    </Page.DataContext>

    <Grid>
        <ListView ItemsSource="{x:Bind ViewModel.Sections}">
            <ListView.ItemTemplate>
                <DataTemplate x:DataType="local:RootViewModel+NestedViewModel">
                    <TextBlock Text="{x:Bind SampleText}"/>
                </DataTemplate>
            </ListView.ItemTemplate>
        </ListView>
    </Grid>
</Page>
```

The app will fail to build and that data type for the template will not be recognized.

**Steps to reproduce the bug**

1. Create a .NET Standard 2.0 project and add the viewmodel code
2. Create a new blank UWP app and reference the other project
3. Paste that XAML code above in `MainPage`
4. Try to build and run

**Expected behavior**
The app should run normally and XAML should recognize the data type and see the properties in the model.

**Screenshots**

![image](https://user-images.githubusercontent.com/10199417/92164883-68ac7a00-ee36-11ea-9f7a-52bed6726e34.png)

**Version Info**

NuGet package version: `n/a`

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
| Yes |  |

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| Insider Build (xxxxx)              |    |
| May 2020 Update (19041)            | Yes   |
| November 2019 Update (18363)       |    |
| May 2019 Update (18362)            |    |
| October 2018 Update (17763)        |    |
| April 2018 Update (17134)          |    |
| Fall Creators Update (16299)       |    |
| Creators Update (15063)            |    |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            | Yes |
| Xbox               |  |
| Surface Hub        |  |
| IoT                |  |

**Comment by BreeceW** [comment]
As pointed out by [reflectronic](https://github.com/reflectronic), this can be accomplished by replacing `x:DataType="local:RootViewModel.NestedViewModel"` with `x:DataType="local:RootViewModel+NestedViewModel"`. The above sample works when this is done.
Hopefully there is a place in the docs where this can be mentioned explicitly.

**Comment by Felix-Dev** [comment]
While the '+' syntax works, the VS XAML editor currently lacks proper support for it: It keeps complaining that the type "local:RootViewModel+NestedViewModel" does not exist:
![image](https://user-images.githubusercontent.com/1398851/92236042-f9c23600-eeb4-11ea-865c-8718025d9172.png)

And as a consequence of that, x:Bind also does not list any of the public APIs of the `NestedViewModel` class in VS (meaning they have to be typed manually and any errors here (such as typos) will only be discovered during compilation).

**Comment by Sergio0694** [comment]
Oh that explains it! I had in fact tried to use `+` already, but assumed it wasn't supported when I saw VS showing those errors.
I'm glad to hear the feature itself is technically working instead, and that it's mostly an issue with the tooling in VS 😄

Should I open this as a feedback in the VS feedback hub and close this issue?

**Comment by Sergio0694** [comment]
Alright I figured out what is causing the issue 🚀

@BreeceW Looks like the XAML compiler fails whenever the target type is defined in a separate assembly. In my case, I have a separate .NET Standard 2.0 project with all my viewmodels, and with that even the `x:DataType="ns:Root+Nested"` syntax fails.

You should be able to repro this locally by just creating a .NET Standard 2.0 project and moving the test viewmodel there.
I assume this _should_ work and it's just a bug in the XAML compiler that causes the type lookup to fail?

Also, would you be able to suggest any possible workarounds for now (other than just moving the types to the app project)?
Thanks! 😊

**Comment by Felix-Dev** [comment]
@Sergio0694 Sorry, just saw your ping now. 

Yes, if I move the class to a separate assembly, I cannot compile any longer as well (previously, I had both the XAML and the ViewModel class in the same assembly).

**Comment by baskren** [comment]
Yet another reason why it was and still is a TERRIBLE idea to not be able to programmatically create DataTemplates.  XAML is not a programming language.

**Comment by Sergio0694** [comment]
Yes, using XAML is clearly a terrible idea because of this one bug.
For the same reason, using C# is also a terrible idea because one time I found a bug in the C# compiler.

This is not the right place to have this conversation, and that comment is completely off topic.
There is a [separate proposal](https://github.com/microsoft/microsoft-ui-xaml/issues/2898) to create data templates in C#, I'd recommend subscribing there if you're interested 🙂

**Comment by baskren** [comment]
Thanks Sergio for being a good sport.  Sorry ... I guess I just needed to vent a bit.

**Comment by Sergio0694** [comment]
Bumping this (see https://github.com/microsoft/microsoft-ui-xaml/discussions/8638).

**Comment by TonyLugg** [comment]
Still now working in .NET 8

**Comment by hippieZhou** [comment]
> Still now working in .NET 8

I also have this issue in my WinUI project

**Comment by AndrewKeepCoding** [comment]
I just confirmed that the `+` syntax works on Microsoft.WindowsAppSDK v1.6.241114003 even though VS throws some errors.

**Comment by AndrewKeepCoding** [comment]
I found another "workaround" that works (if `NestedViewModel` is not `sealed`).

Just create a class that inherits from `NestedViewModel`:
```cs
public class MyNestedViewModel : RootViewModel.NestedViewModel {}
```

then use it on XAML:
```xaml
<DataTemplate x:DataType="local:MyNestedViewModel">
    <TextBlock Text="{x:Bind SampleText}"/>
</DataTemplate>
```

**Comment by sinatrocious** [comment]
> Just create a class that inherits from `NestedViewModel`:
> 
> public class MyNestedViewModel : RootViewModel.NestedViewModel {}

I am getting warning and it bindings to properties aren't working at all:

> Microsoft.Maui.Controls.Xaml.Diagnostics.BindingDiagnostics: Warning: Mismatch between the specified x:DataType (MyNestedViewModel ) and the current binding context (RootViewModel+NestedViewMode).

**Comment by AndrewKeepCoding** [comment]
@sinatrocious You are right. That doesn't work and I removed my earlier comment.
It looks like we need to wait until this is properly addressed.😅

If you absolutely need a temporary hack, check out this [repo](https://github.com/AndrewKeepCoding/WinUIDemos/tree/msbuild-tasks).

It uses a custom MSBuild task that edits the generated *.g.cs* files and replaces code like `RootViewModel+NestedViewMode` with `RootViewModel.NestedViewMode`.

I was just experimenting to see if it worked. While it seems that it does, I don't recommend it. Custom MSBuild tasks seems to have some file locking [issue](https://jaylee.org/archive/2019/01/18/chasing-an-msbuild-task-lock.html) that I couldn't workaround.

**Comment by Sergio0694** [comment]
Closing this, I merged the PR with a fix for this a while back (will probably be in 2.4.x or 2.5.x I assume).
